core/src/machine/step.rs

```diff
@@ -577,6 +528,25 @@ impl Machine {
                     self.cycle_counter,
                 );
             }
+        }
+
+        // Unwrap the action_taken `Option` to check, if it was set when we get a branch instruction
+        let action_taken = match instr.clocks_taken {
+            Some(_) => {
+                match action_taken {
+                    Some(t) => t,
+                    None => {
+                        terminate!(
+                            "action_taken not set for branch instruction {:?} in position: \
+                                {} after: {} cycles!",
+                            instr,
+                            instr_start,
+                            self.cycle_counter,
+                        );
+                    },
+                }
+            },
+            None => false,
```

Comment:
You could also write it like this:

```rust
       let action_taken = match (instr.clocks_taken, action_taken) {
    (Some(_), Some(t)) => t,
    (Some(_), None) => {
        terminate!(
            "action_taken not set for branch instruction {:?} in position: \
                {} after: {} cycles!",
            instr,
            instr_start,
            self.cycle_counter,
        );
    }
    (None, Some(_)) => {
        // We should error here as well, right?
    }
    (None, None) => false,
}
```

While writing that I noticed this fourth option: a non branch instruction setting the flag. This should also error, I guess.

But yeah, as a general tip: it's really neat to sometimes match over a pair of two values.